### ISSUE TITLE
Documentation on expected errors payload?

### ISSUE THREAD

**Issue by andrewhavens** [ember-cli/ember-ajax#109]
I'm running into an issue where my errors response is not as I expect, but I can't find any documentation on what the expected format is supposed to be.

Currently, my errors response is formatted like this (JSONAPI spec, I guess):

```
{
  errors: [
    "Error message"
  ]
}
```

But the error response from my ajax request makes it look like an array of objects.

<img width="180" alt="screen shot 2016-05-11 at 4 28 36 pm" src="https://cloud.githubusercontent.com/assets/446392/15200294/45276e4a-179a-11e6-8241-02ce83770b38.png">

I really just don't know how I am supposed to work with the errors object because I can't find any documentation about it.

```js
this.get('ajax').post('endpoint', payload).then(success).catch(function(e){
  console.error("errors", e.errors); // looks like an array of objects
  console.error("messages", e.errors.messages); // undefined
  console.error(e.errors.join('. ')); // [Object object].
});
``

**Comment by andrewhavens** [comment]
I think the version of ember-ajax I am using is 2.0.1, based on my package.json:

```
  "devDependencies": {
    "ember-ajax": "2.0.1",
```

**Comment by alexlafroscia** [comment]
Hey @andrewhavens! Thanks for bringing this to my attention.  As far as I know, `errors` should be an array of error objects.  I don't think there was anything that formally tied it back to the JSON:API spec, but if you check it out, it does reference the fact that the `errors` key would correspond to an array of objects:

http://jsonapi.org/format/#error-objects

Sorry that the documentation doesn't cover that at all. I'll write up some better docs and get the published as part of the Wiki, and link to that from the README.

**Comment by wasabigeek** [comment]
@alexlafroscia I'm trying to display the errors returned by Django Rest Framework (DRF) e.g. 
```
{
    "email": [
        "Enter a valid email address."
    ]
}
```
I presume this is in error.errors.details, but like andrew's example all I seem to get is `[Object object]`. How can I display the error from DRF?

**Comment by alexlafroscia** [comment]
@wasabigeek I'll make sure to cover that in my documentation.  The best approach would be to make your own sub-class of the `ajax` service (so, make your own service, import the `AjaxService` and extend from that) and then overwrite the `normalizeErrorResponse` method to do something like this (no guarantees that this exact code would work, though; this is just off the top of my head)

```js
normalizeErrorResponse(status, headers, payload) {
  if (Ember.isArray(payload.errors)) {
    return payload.errors.map(function(error) => {
      return { detail: error };
    });
  }
}
```

This _should_ set things up so that the `AjaxError` class receives the format that it's expecting.

If there's any suggestions on how we could make this easier, I would love to hear them! The idea is for `ember-ajax` to be extensible for exactly this purpose, and I realize the docs aren't great right now.  I'm working on it!

**Comment by bakura10** [comment]
I'm having a similar problem. My API returns error this way:

```json
{
  "message": "foo",
  "another_key": "bar"
}
```

Unfortunately, none of those errors can be retrieved in the error object. I don't think this low-level component should go that far into normalization, or at least it should allows us to retrieve the original payload without having to create a custom service :).

**Comment by alexlafroscia** [comment]
I'm not sure about that one. At some point we have to make some assumptions, that's just the "convention over configuration" nature of Ember. I don't see subclassing the service to retrieve your custom error format as being a huge deal. Since Ember Data seems to be leaning pretty heavily on the `json:api` spec, we use that as our standard as well. 

Ideally, how would you like to be able to access the original error, without needing to subclass anything? Should the `AjaxError` object that the promise is rejected with contain the entire payload from the server?

**Comment by bakura10** [comment]
Yeah, convention is good, but maybe not at the expense on not even being able to retrieve an error payload :). Something as simple as "errors.originalPayload" would do the trick. The confusing prat for me was that on success, I can retrieve the payload as usual, even if it does not follow any known standard (my API is not JSON API compliant), while it's not possible for the errors.

**Comment by billpull** [comment]
From following the docs I created a sub class of the ajax service in `/app/services/ajax.js` and  added a  `normalizeErrorResponse` method. However in tests when I set a breakpoint in `handleResponse` it still has the reference to the default and ignores my service.

```
import Ember from 'ember';
import AjaxService from 'ember-ajax/services/ajax';

export default AjaxService.extend({
  normalizeErrorResponse(status, headers, payload) {
    if (payload.errorCode && payload.message) {
      return [{ title: payload.errorCode, detail: payload.message }];
    }

     return this._super(status, headers, payload);
   }
});
```

**Comment by chrism** [comment]
I've just found myself with the same issue as @bakura10 and think it is unfortunate that there is no way to retrieve the original error payload without sub-classing the `ajax` service itself, something that was pretty straightforward before with `ic-ajax`.

As an example, if I use several different 3rd party APIs and they each have their own error response format it quickly becomes complex to handle each individual format in `normalizeErrorResponse`.

Don’t get me wrong, i'm a big fan of convention over configuration I just wish that in this situation there is a straightforward escape hatch to use as API response formats are often something that are outside our control.

Besides this one frustration I’ve found `ember-ajax` to be a really great addition to the ecosystem and your hard work on it is really appreciated.

**Comment by alexlafroscia** [comment]
Sorry I haven't gotten to actually helping here, I'm in the middle of a move to SF and have had no OSS time lately.

Let me think about how to expose both the original and formatted error payloads.  

>  if I use several different 3rd party APIs and they each have their own error response format

I think the approach we had in mind here would be to use three separate `ajax` service sub-classes, where each has their own `normalizeErrorResponse` defined.  This way they all have their own error formatter, namespace, host (if necessary) etc.  I understand that that may not be how you want to work with the API, though.  Does doing that seem like too much?  To be honest, I don't have much experience as a _consumer_ of `ember-ajax`, despite maintaining it 😄 

Also:

> Besides this one frustration I’ve found ember-ajax to be a really great addition to the ecosystem and your hard work on it is really appreciated.

Thank you! I really appreciate that 😀

**Comment by chrism** [comment]
Hi @alexlafroscia, hope the move went OK! Sorry for my slow reply but I wanted to give you a clear example so had to free up a bit of time to make an Ember Twiddle working with Ember-Ajax.

[Ember-Ajax Non Standard Error Responses](https://canary.ember-twiddle.com/976ac62f3122b4dbcfe468fe05650087?openFiles=services.ajax.js)

> I think the approach we had in mind here would be to use three separate ajax service sub-classes, where each has their own `normalizeErrorResponse` defined. This way they all have their own error formatter, namespace, host (if necessary) etc. I understand that that may not be how you want to work with the API, though.

In my example it shows that assuming your server responds correctly everything just works great, but if it doesn't it becomes non-trivial to handle the error in a similar way.

You can see that it requires both the `normalizeErrorResponse` hook to format the json response into a JSON API friendly format (taken from this thread) and `isSuccess` hook to trigger an error (illustrated in the documentation) for an accessible error package in the promise rejection handler.

Unless in doing this I have missed something obviously simpler then by this point, for me at least, it has become a bit of a sledgehammer to crack a nut. Also, multiply this boilerplate when sub-classing each time depending on the specific error response.

Personally my use case for these types of ajax calls are when I want something specific outside of Ember Data. In the end I found it easier to use `handleResponse` in my Ember Data adapter to format the error and to use `urlForQueryRecord` to customize the endpoint — removing the need entirely for ember-ajax in this instance.

I completely understand it is all a question of balance but wanted to explain fully the circumstances and reasoning behind my point.

**Comment by jackmatt2** [comment]
Hit this issue also, was unsure why my payload was not visible which I can clearly see it in the JSON response.  Access to the original "unparsed" value would be great.

**Comment by Ynnni** [comment]
Just look at this code:
```js
request(url, options) {
  const hash = this.options(url, options);
  return new Promise((resolve, reject) => {
    this.raw(url, hash)
      .then(({ response }) => {
        resolve(response);
      })
      .catch(({ response }) => {
        reject(response);
      });
  }, `ember-ajax: ${hash.type} ${hash.url} response`);
}
```
and this:
```js
hash.error = (jqXHR, textStatus, errorThrown) => {
  const payload = this.parseErrorResponse(jqXHR.responseText) || errorThrown;
  let response;

  if (errorThrown instanceof Error) {
    response = errorThrown;
  } else if (textStatus === 'timeout') {
    response = new TimeoutError();
  } else if (textStatus === 'abort') {
    response = new AbortError();
  } else {
    response = this.handleResponse(
       jqXHR.status,
       parseResponseHeaders(jqXHR.getAllResponseHeaders()),
       payload,
       requestData
    );
  }

  pendingRequestCount--;

  run.join(null, reject, { payload, textStatus, jqXHR, errorThrown, response });
};

```

Addon use http://es6-features.org/#ParameterContextMatching and ignore all parameters sent from this
```js
run.join(null, reject, { payload, textStatus, jqXHR, errorThrown, response });
```
to this catch
```js
.catch(({ response }) => {
  reject(response);
});
```
except response

**Comment by Ynnni** [comment]
Can i create pull request which do something like this?
```js
.catch(object) => {
  reject(object);
});
```

**Comment by taras** [comment]
@Ynnni is the intention to allow `request` method to receive additional response data?

IOW, to make `request` more like `raw`?

**Comment by taras** [comment]
@chrism it looks like your code should probably be part of Ember Ajax. In a situation when `payload.error` is returned, we should turn push it into `payload.errors`. What do you think?

**Comment by chrism** [comment]
> @chrism it looks like your code should probably be part of Ember Ajax. In a situation when payload.error is returned, we should turn push it into payload.errors. What do you think?

Hi @taras thanks for replying on this, sorry it has taken me so long to get back to you.

If I understand your proposal correctly then this would certainly have been a good solution for my particular use case.

Though the broader concern (that by removing any possible access to the raw payload means that any non-standard approach not covered by the library becomes very difficult to incorporate) remains.

Even with this additional support I'm pretty convinced that sooner or later there will be a scenario where the `errors.originalPayload` (or similar) that @bakura10 proposed would be still be required.

But thats just, like, my opinion :smile:

Thanks for all your work in the Ember community 👍

**Comment by taras** [comment]
@chrism @alexlafroscia and I spent some time looking at this problem and it turns out this part of this problem resolves around `ember-ajax` not normalizing errors correctly. We had scenarios when errors were arrays, objects or strings. 

PR #118 is going to address this problem which also provides `originalPayload` in the meta property of the error. This should give people enough on an escape valve to handle most of their use cases.

**Comment by chrism** [comment]
In which case that sounds perfect to me 👍
